File: 2025/golang/day6/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
	"strconv"
)


// Perform the desired operation
func performOperationStr(numbers []string, operation string) int {
	result := 0
	if operation == "*" {
		result = 1 
	} else if operation == "+" {
		result = 0
	}
	nums := []string{}
	numsInt := []int{}
	for i := 0; i < len(numbers[0]); i++ {
		nums = append(nums, "")
	}

	for i := 0; i < len(numbers[0]); i++ {
		for j := range numbers {
			nums[i] = nums[i] + string(numbers[j][i])
		}
	}

	for _, num := range nums {
		int, err := strconv.Atoi(strings.ReplaceAll(num, " ", ""))
		if err != nil {
			panic(err)
		}
		numsInt = append(numsInt, int)
	}

	for _, num := range numsInt {
		if operation == "+" {
			result += num
		} else if operation == "*" {
			result *= num
		}
	}

	return result
}

// Get indexes of spaces in each line, and count how many times each index has a space
func getSpacesIndexes(lines []string) map[int]int {
	spaces := make(map[int]int)
	for _, line := range lines {
		for i := 0; i < len(line); i++ {
			if string(line[i]) == " " {
			spaces[i]++
			}
		}
	}

	return spaces
}

// Get indexes of spaces that are common to all lines, and also include -1 and the last index of the line
func getCommonSpaces(spaces *map[int]int, maxSeparator, maxIdx int) []int {
	commonSpaces := []int{}
	for key, val := range *spaces {
		if val == maxSeparator {
			commonSpaces = append(commonSpaces, key)
		}
	}

	commonSpaces = append(commonSpaces, -1)
	commonSpaces = append(commonSpaces, maxIdx)
	sort.Ints(commonSpaces)
	return commonSpaces
}

// Find the maximum value in the map of spaces
// This is used so that we can find the common spaces
// NOTE: This should be the same as the number of lines in the input, 
// because the common spaces are the spaces that are present in all lines
func findMaxSeparator(spaces *map[int]int) int {
	maxSeparator := 0
	for _, val := range *spaces {
		if val > maxSeparator {
			maxSeparator = val
		}
	}
	return maxSeparator
}

// Get operations from last line
func getOperations(lines []string) []string {
	ops := []string{}
	for i := 0; i < len(lines[len(lines)-1]); i++ {
		if lines[len(lines)-1][i] != ' ' {
			ops = append(ops, string(lines[len(lines)-1][i]))
		}
	}
	return ops
}

// Just construct the matrix of numbers based on the common spaces found in the lines
// aligned in columns
func constructMatrix(lines []string, commonSpaces []int) [][]string {
	matrix := [][]string{}
	for i := 0; i < len(commonSpaces)-1; i++ {
		row := []string{}
		for _, line := range lines {
			row = append(row, string(line[commonSpaces[i]+1:commonSpaces[i+1]]))

		}
		matrix = append(matrix, row)
	}
	return matrix
}

// find the longest line in the input to make sure we 
// include trailing spaces in each line to make them all the same length
func findLongestLine(lines []string) int {
	longestLine := 0
	for _, line := range lines {
		if len(line) > longestLine {
			longestLine = len(line)
		}
	}
	return longestLine
}

// Include trailing spaces in each line to make them all the same length
// because sometimes the last number would be x in one line and yy in another line, 
// and we need to make sure they are all the same length because the digits in the 
// operations are aligned by their position in the line.
func includeTrailingSpaces(lines *[]string, longestLine int) {
	for i := 0; i < len(*lines); i++ {
		if len((*lines)[i]) < longestLine {
			(*lines)[i] = (*lines)[i] + strings.Repeat(" ", longestLine-len((*lines)[i]))
		}
	}
}

func solveDay6Part2(input string) {
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	total := 0
	scanner := bufio.NewScanner(file)
	lines := []string{}

	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	ops := getOperations(lines)
	lines = lines[:len(lines)-1] // exclude operations
	longestLine := findLongestLine(lines)
	includeTrailingSpaces(&lines, longestLine)
	spaces := getSpacesIndexes(lines)
	maxSeparator := findMaxSeparator(&spaces)
	commonSpaces := getCommonSpaces(&spaces, maxSeparator, len(lines[0]))
	matrix := constructMatrix(lines, commonSpaces)

	for i := 0; i < len(matrix); i++ {
		total += performOperationStr(matrix[i], ops[i])
	}

	fmt.Println("Total: ", total)
}

func main() {
	// input := "input_test.txt" // Expected output: 3263827
	input := "input.txt" // 10756006415204
	solveDay6Part2(input)
}

File: 2025/cpp/day6/part2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <algorithm>
#include <string>
#include <map>


std::string remove_spaces(const std::string& str) {
    std::string result;
    for (char c : str) {
        if (c != ' ') {
            result += c;
        }
    }
    return result;
}

long long perform_operation_str(std::vector<std::string>& numbers, std::string operation) {
    using namespace std;

    long long result = 0;
    if (operation == "*") {
        result = 1;
    }

    vector<string> nums;
    vector<long long> nums_ll;
    for (size_t i = 0; i < numbers[0].length(); i++) {
        nums.push_back("");
    }

    for (size_t i = 0; i < numbers[0].length(); i++) {
        for (size_t j = 0; j < numbers.size(); j++) {
            nums[i] += numbers[j][i];
        }
    }

    for (auto num : nums) {
        string num_wo_spaces = remove_spaces(num);
        if (num_wo_spaces.empty()) {
            continue;
        }
        nums_ll.push_back(stoll(num_wo_spaces));
    }

    for (auto num : nums_ll) {
        if (operation == "+") {
            result += num;
        } else if (operation == "*") {
            result *= num;
        }
    }

    return result;
}


std::map<long long, int> get_space_indexes(std::vector<std::string> lines) {
    using namespace std;

    map<long long,int> spaces = map<long long, int>();

    for (auto line : lines) {
        for (int i = 0; i < line.length(); i++) {
            if (line[i] == ' ') {
                spaces[i]++;
            }
        }
    }

    return spaces;
}

std::vector<int> get_common_spaces(const std::map<long long, int>& spaces, int max_separator, int max_id) {
    using namespace std;

    vector<int> common_spaces;
    for (const auto& space : spaces) {
        if (space.second == max_separator) {
            common_spaces.push_back(space.first);
        }
    }

    common_spaces.push_back(-1);
    common_spaces.push_back(max_id);
    sort(common_spaces.begin(), common_spaces.end());

    return common_spaces;
}

int find_max_separator(const std::map<long long, int>& spaces) {
    int max_separator = 0;
    for (const auto& space : spaces) {
        if (space.second > max_separator) {
            max_separator = space.second;
        }
    }
    return max_separator;
}

std::vector<std::string> get_operations(const std::vector<std::string>& lines) {
    using namespace std;

    vector<string> operations;
    for (size_t i = 0; i < lines[lines.size() - 1].length(); i++) {
        if (lines[lines.size()-1][i] != ' ') {
            operations.push_back(string(1, lines[lines.size()-1][i]));
        }
    }

    return operations;
}

std::vector<std::vector<std::string>> construct_matrix(
    std::vector<std::string> lines,
    std::vector<int> common_spaces
) {
    using namespace std;

    vector<vector<string>> matrix;
    for (size_t i = 0; i < common_spaces.size() - 1; i++) {
        vector<string> row;
        for (auto line : lines) {
          row.push_back(line.substr(
              common_spaces[i] + 1,
              common_spaces[i + 1] - common_spaces[i] - 1
          ));
        }

        matrix.push_back(row);
    }

    return matrix;
}


int find_longest_line(const std::vector<std::string>& lines) {
    int max_length = 0;
    for (const auto& line : lines) {
        if (line.length() > max_length) {
            max_length = line.length();
        }
    }
    return max_length;
}

std::vector<std::string> split(const std::string& str, char delimiter) {
    using namespace std;

    vector<string> tokens;
    string token;
    istringstream tokenStream(str);
    while (getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}

void include_trailing_spaces(std::vector<std::string>& lines, int max_length) {
    for (auto& line : lines) {
        while (line.length() < max_length) {
            line += " ";
        }
    }
}

void solve_day6_part2(std::string input_file) {
    using namespace std;

    long long total = 0;
    ifstream input(input_file);
    string line;
    vector<string> lines;

    while (getline(input, line)) {
        lines.push_back(line);
    }

    vector<string> operations = get_operations(lines);
    lines = vector<string>(lines.begin(), lines.end() - 1);
    int longest_line = find_longest_line(lines);
    include_trailing_spaces(lines, longest_line);
    map<long long, int> spaces = get_space_indexes(lines);
    int max_separator = find_max_separator(spaces);
    vector<int> common_spaces = get_common_spaces(spaces, max_separator, lines[0].length());
    vector<vector<string>> matrix = construct_matrix(lines, common_spaces);

    for (size_t i = 0; i < matrix.size(); i++) {
        total += perform_operation_str(matrix[i], operations[i]);
    }

    cout << "Total: " << total << endl;
}


int main() {
    // std::string input_file = "input_test.txt"; // Expected output: 3263827 
    std::string input_file = "input.txt"; // 10756006415204
    solve_day6_part2(input_file);
}
